bindings/mqtt/mqtt.go

```diff
@@ -243,58 +243,96 @@ func (m *MQTT) handleMessage(ctx context.Context, handler bindings.Handler, mqtt
 }
 
 func (m *MQTT) Read(ctx context.Context, handler bindings.Handler) error {
-	// reset synchronization
-	if m.consumer != nil {
-		m.logger.Warnf("re-initializing the subscriber")
-		m.consumer.Disconnect(5)
-		m.consumer = nil
+	// mqtt broker allows only one connection at a given time from a clientID
+	topicHash := sha256.Sum256([]byte(m.metadata.topic))
+	consumerClientID := fmt.Sprintf("%s-consumer-%x", m.metadata.clientID, topicHash[0:4])
+
+	// Subscribe to the topics on the OnConnectHandler callback
+	readyCh := make(chan error)
+	onConn := func(client mqtt.Client) {
+		token := client.Subscribe(m.metadata.topic, m.metadata.qos, m.onMessage(ctx, handler))
+		ok := token.WaitTimeout(defaultWait)
+		tokenErr := token.Error()
+		// In case of timeout, ok will be false and there won't be an error in the token
+		if tokenErr == nil && !ok {
+			tokenErr = context.DeadlineExceeded
+		}
+		readyCh <- tokenErr
 	}
 
-	// mqtt broker allows only one connection at a given time from a clientID.
-	consumerClientID := fmt.Sprintf("%s-consumer", m.metadata.clientID)
-	c, err := m.connect(consumerClientID)
+	m.logger.Infof("mqtt subscribing to topic '%s' with consumer client ID '%s'", m.metadata.topic, consumerClientID)
+	client, err := m.connect(consumerClientID, onConn)
 	if err != nil {
 		return err
 	}
-	m.consumer = c
+	// m.connect returns after the connection has been established, now we need to wait for the subscription to be up
+	select {
+	case <-time.After(defaultWait):
```

Comment:
if I understand correctly, I would recommend that it would be better to use time.NewTimer instead of time.After, to avoid memory leak scenarios. (https://pkg.go.dev/time#After)

Reply:
I will change it, but the risk of leaks is not a problem here. Per the docs:

> The underlying Timer is not recovered by the garbage collector until the timer fires.

Which does indeed mean there's a "leak" but only "until the timer fires". In this case, the timer is short-lived (`defaultWait` is 30s), so it's not a big concern here.